core/lib/via_musig2/src/lib.rs

```diff
@@ -0,0 +1,253 @@
+use std::fmt;
+
+use musig2::{
+    verify_single, CompactSignature, FirstRound, KeyAggContext, PartialSignature, PubNonce,
+    SecNonceSpices, SecondRound,
+};
+use secp256k1_musig2::{PublicKey, Secp256k1, SecretKey};
+
+#[derive(Debug)]
+pub enum MusigError {
+    Musig2Error(String),
+    InvalidSignerIndex,
+    MissingNonces,
+    MissingPartialSignatures,
+    InvalidState(String),
+}
+
+impl fmt::Display for MusigError {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            MusigError::Musig2Error(e) => write!(f, "MuSig2 error: {}", e),
+            MusigError::InvalidSignerIndex => write!(f, "Invalid signer index"),
+            MusigError::MissingNonces => write!(f, "Missing required nonces"),
+            MusigError::MissingPartialSignatures => {
+                write!(f, "Missing required partial signatures")
+            }
+            MusigError::InvalidState(s) => write!(f, "Invalid state: {}", s),
+        }
+    }
+}
+
+impl std::error::Error for MusigError {}
+
+/// Represents a single signer in the MuSig2 protocol
+pub struct Signer {
+    secret_key: SecretKey,
+    _public_key: PublicKey,
+    signer_index: usize,
+    key_agg_ctx: KeyAggContext,
+    first_round: Option<FirstRound>,
+    second_round: Option<SecondRound<Vec<u8>>>,
+    message: Vec<u8>,
+    nonce_submitted: bool,
+    partial_sig_submitted: bool,
+}
+
+impl Signer {
+    /// Create a new signer with the given secret key and index
+    pub fn new(
+        secret_key: SecretKey,
+        signer_index: usize,
+        all_pubkeys: Vec<PublicKey>,
+    ) -> Result<Self, MusigError> {
+        let secp = Secp256k1::new();
+        let public_key = PublicKey::from_secret_key(&secp, &secret_key);
```

Comment:
Maybe we can check here that the computed `public_key` is indeed passed in `all_pubkeys` at `signer_index`?

Reply:
I think it's better to doing this in withdrawal service, during the initialization of service to prevent doing the same check for every signing processes

wdyt?